Refactor vectors

MichielStock opened the pull request.

MichielStock committed:
:workinprogress: refactor

MichielStock committed:
:hammer: rework vectors

MichielStock committed:
:white_check_mark: tests for vectors

MichielStock committed:
:wrench: removed distr from graded

MichielStock committed:
:work_in_progess: aggregation

cvigilv committed:
docs: add intro to HDC and Kanerva's Mexico Dollar

cvigilv committed:
fix: add missing dependencies for docs generation

MichielStock committed:
:wrench: simplify types

MichielStock committed:
:work_in_progress: operations

MichielStock committed:
:tada: operators with tests!

MichielStock committed:
upgrade manifest

cvigilv committed:
docs(wip): package tutorial

MichielStock committed:
:tada: implemented the approx op

MichielStock committed:
:tada: similarities

cvigilv commented:
@MichielStock can you add me as reviewer of the PR, that way I can do it here and discuss possible changes before merging

cvigilv committed:
docs: continue working on documentation

- Remove `build` folder
- Improve "Introduction" page
- Add "What's the dollar of Mexico?" page
- Update Manifest.toml and Project.toml

cvigilv commented on src/HyperdimensionalComputing.jl:
I think we could organizing the module file as follows:
```julia
module HyperdimensionalComputing

using ...

include("vector.jl")
export AbstractHV,
  BinaryHV, 
  BipolarHV,
  GradedBipolarHV, 
  RealHV, 
  GradedHV, 
  TernaryHV

include("operations.jl")
export aggregate, bind, shift!, shift, ρ, ρ!

...
```

Why this way: this provides an entry point to the package for new contibutors. Think of it as a glossary for the package.

cvigilv reviewed (commented):
Code review o'clock!

MichielStock committed:
:tada: permute

MichielStock committed:
:work_in_progress: level encoding

MichielStock commented on src/inference.jl:
This needs some tweaking. There should be a good default for each HV, but one should be able to override it. You can remove the export.

MichielStock commented on src/operations.jl:
```diff
@@ -11,43 +11,45 @@ operations.jl; This file implements operations that can be done on hypervectors
 Maps a graded number in [0, 1] to the [-1, 1] interval.
 """
-grad2bipol(x::Number) = 2x - one(x)
+grad2bipol(x::Real) = 2x - one(x)
```
I think this is only used internally.

MichielStock commented on src/operations.jl:
```diff
@@ -127,10 +184,21 @@ function aggregate!(r::AbstractHDV, hdvs, weights; clear=true, norm=false)
 # BINDING
 # -------
 
+Base.bind(hv1::BinaryHV, hv2::BinaryHV) = BinaryHV(hv1.v .⊻ hv2.v)
```
`BinaryHV` and `BipolarHV` always work a bit different because they use bitvectors internally. I have written a trait today to solve this. I need to take a look.

cvigilv commented on src/types.jl:
```diff
@@ -0,0 +1,148 @@
+Base.getindex(hv::BipolarHV, i) = hv.v[i] ? 1 : -1
+Base.sum(hv::BipolarHV) = 2sum(hv.v) - length(hv.v)
+LinearAlgebra.norm(hv::BipolarHV) = sqrt(length(hv))
+
```
I have something already implemented for plotting function. Once we release the v1.0 we could add this as a package extension.

cvigilv commented on src/types.jl:
```diff
@@ -0,0 +1,148 @@
+
+# Gives an empty Vector (filled with neutral elelment) that
+# the `hv::AbstractHV` type uses.
+empty_vector(hv::AbstractHV) = zero(hv.v)
```
I think this could be exported. It looks like a good way of initializing prototype hypervectors.

cvigilv committed:
chore: merge branch 'main'

cvigilv committed:
docs: remove Manifest.jl

MichielStock committed:
:broom: aggregate => bundle + remove

cvigilv committed:
chore: Merge branch 'main'

cvigilv committed:
chore: merge pull request #26 from cvigilv/refactor_vectors

refactor: vectors
